diff --git a/configure.in b/configure.in
index 81acf76..a81525f 100644
--- a/configure.in
+++ b/configure.in
@@ -298,9 +298,9 @@ if test "$OCAMLFIND" = "no" -o "$OCAMLGRAPH_CUSTOM" = "yes"; then
 
 # check if any ocamlgraph is installed in the right place
   AC_CHECK_FILE($OCAMLGRAPH_HOME/graph.$OBJ_SUFFIX,
-              OCAMLGRAPH_EXISTS="yes" OCAMLGRAPH_INCLUDE="-I +ocamlgraph")
+              OCAMLGRAPH_EXISTS="yes" OCAMLGRAPH_INCLUDE="-I $OCAMLGRAPH_HOME")
 
-  if test "$OCAMLGRAPH_EXISTS"="no" -a "$OCAMLGRAPH_CUSTOM" = "yes"; then
+  if test "$OCAMLGRAPH_EXISTS" = "no" -a "$OCAMLGRAPH_CUSTOM" = "yes"; then
     AC_MSG_ERROR(
      [unable to find OCamlGraph in custom location $OCAMLGRAPH_HOME.
       Please give an appropriate value, or use default OCamlGraph library]);
@@ -341,7 +341,7 @@ if test "$OCAMLGRAPH_EXISTS" = "yes"; then
     1.8.1) ocamlgraph_error;;
     1.8.2) ocamlgraph_error;;
     1.8.3) ocamlgraph_error;;
-    1.8.4) AC_MSG_NOTICE([OcamlGraph $OCAMLGRAPH_VERSION found: great!]);;
+    1.8.4*) AC_MSG_NOTICE([OcamlGraph $OCAMLGRAPH_VERSION found: great!]);;
     1.8.*) AC_MSG_NOTICE(
       [OcamlGraph $OCAMLGRAPH_VERSION > 1.8.4 found: should be compatible, but no warranty. Use it at your own risk!]);;
     *) ocamlgraph_error;;
