diff --git a/configure.in b/configure.in
index 1bb224c..348898b 100644
--- a/configure.in
+++ b/configure.in
@@ -300,7 +300,7 @@ else
       FRAMACVERSION=`$FRAMAC -version | sed -n -e 's|Version: *\(.*\)$|\1|p' `
       AC_MSG_RESULT($FRAMACVERSION)
       case $FRAMACVERSION in
-        Fluorine-20130401) 
+        Fluorine-*)
            FRAMAC=yes
            ;;
         *) FRAMACMSG="you need Frama-C version Fluorine-20130401"
