diff --git a/java/java_interp.ml b/java/java_interp.ml
index 894c710..603d555 100644
--- a/java/java_interp.ml
+++ b/java/java_interp.ml
@@ -2170,7 +2170,7 @@ let tr_axiomatic_axiom acc d =
     | Atype _s -> acc
     | Adecl(_fi, _b) -> acc
 
-let tr_axiomatic id l acc =
+let tr_axiomatic acc (id, l) =
   Java_options.lprintf "translating axiomatic %s@." id;
   let acc1 = List.fold_left tr_axiomatic_decl [] l in
   let acc2 = List.fold_left tr_axiomatic_axiom acc1 l in
diff --git a/java/java_main.ml b/java/java_main.ml
index da8b48b..5719e56 100644
--- a/java/java_main.ml
+++ b/java/java_main.ml
@@ -201,8 +201,7 @@ let main () =
       decls
   in
   let decls =
-    Hashtbl.fold Java_interp.tr_axiomatic Java_typing.axiomatics_table
-      decls
+    List.fold_left Java_interp.tr_axiomatic decls !Java_typing.axiomatics_table
   in
   (* production phase 1.5 : generation of Jessie lemmas *)
   let lemmas =
diff --git a/java/java_typing.ml b/java/java_typing.ml
index 10f83ab..b5f8db7 100644
--- a/java/java_typing.ml
+++ b/java/java_typing.ml
@@ -579,7 +579,7 @@ type axiomatic_defs =
 
 let logics_env = Hashtbl.create 97
 let logic_defs_table = Hashtbl.create 97
-let axiomatics_table = Hashtbl.create 97
+let axiomatics_table = ref []
 
 let builtin_logic_symbols = ref []
 
@@ -4194,7 +4194,7 @@ let rec type_decl_aux ~in_axiomatic package_env type_env acc d =
 	let l = List.fold_left
 	  (type_decl_aux ~in_axiomatic:true package_env type_env) [] l
 	in
-	Hashtbl.add axiomatics_table id l;
+	axiomatics_table := !axiomatics_table @ [(id, l)];
 	acc
     | JPTimport(PolyTheoryId(id_list,_params)) ->
         let loc = fst (List.hd id_list) in
@@ -4212,7 +4212,7 @@ let rec type_decl_aux ~in_axiomatic package_env type_env acc d =
 	  (type_decl_aux ~in_axiomatic:true package_env type_env) [] body
 	in
         let id = snd (List.hd id_list) in
-	Hashtbl.add axiomatics_table id l;
+	axiomatics_table := !axiomatics_table @ [(id, l)];
 	acc
 
 
diff --git a/java/java_typing.mli b/java/java_typing.mli
index 91bd7e6..8656081 100644
--- a/java/java_typing.mli
+++ b/java/java_typing.mli
@@ -107,7 +107,7 @@ type axiomatic_defs =
   | Atype of string
 
 val logic_types_table : (string, Java_env.logic_type_info) Hashtbl.t
-val axiomatics_table : (string, axiomatic_defs list) Hashtbl.t
+val axiomatics_table : (string * axiomatic_defs list) list ref
 
 val builtin_logic_symbols : Java_env.java_logic_info list ref
 
