diff --git a/lib/why3/jessie3theories.why b/lib/why3/jessie3theories.why
index a00e3a2..0a91e0c 100644
--- a/lib/why3/jessie3theories.why
+++ b/lib/why3/jessie3theories.why
@@ -27,7 +27,7 @@ function offset_min (alloc_table 't) (pointer 't) : int
 (* shortcuts *)
 
 predicate valid (a:alloc_table 't) (p:pointer 't) =
-  offset_min a p <= 0 /\ offset_max a p >= 0
+  offset_min a p <= 0 /\ offset_max a p >= -1
 
 predicate same_block (p: pointer 't) (q:pointer 't) =
   base_block(p) = base_block(q)
