From da875ab62ba5aa349143bdec4667b0d3398e0849 Mon Sep 17 00:00:00 2001
From: Mikhail Mandrykin <mandrykin@ispras.ru>
Date: Fri, 10 Jan 2014 15:58:18 +0400
Subject: [PATCH 3/6] Fixed builtin bool type in Why3 output

---
 jc/output.ml |    1 +
 1 file changed, 1 insertion(+)

diff --git a/jc/output.ml b/jc/output.ml
index 965d5cc..ffe25ca 100644
--- a/jc/output.ml
+++ b/jc/output.ml
@@ -581,6 +581,7 @@ let logic_type_name t =
   if !why3syntax then
     match t.logic_type_name with
       | "unit" -> "()"
+      | "bool" -> "Bool.bool"
       | s -> why3ident s
   else
     t.logic_type_name
-- 
1.7.9.5

